src/components/views/messages/TextualBody.js

```diff
@@ -162,6 +168,53 @@ module.exports = React.createClass({
         }
     },
 
+    pillifyLinks: function(nodes) {
+        const MemberAvatar = sdk.getComponent('avatars.MemberAvatar');
+        const RoomAvatar = sdk.getComponent('avatars.RoomAvatar');
+        for (let i = 0; i < nodes.length; i++) {
+            const node = nodes[i];
+            if (node.tagName === "A" && node.getAttribute("href")) {
+                const href = node.getAttribute("href");
+                // HtmlUtils transforms `matrix.to` links to local links, so match against
+                // user or room app links.
+                const match = /^#\/(user|room)\/(.*)$/.exec(href);
+                if (match) {
+                    let avatar;
+                    let roomId;
+                    let room;
+                    let member;
+                    switch (match[1]) {
+                        case "user":
+                            roomId = this.props.mxEvent.getRoomId();
+                            room = MatrixClientPeg.get().getRoom(roomId);
+                            member = room.getMember(match[2]) ||
+                                new RoomMember(null, match[2]);
+                            avatar = <MemberAvatar member={member} width={16} height={16} name={match[2]}/>;
+                        break;
+                        case "room":
+                            room = match[2][0] === '#' ?
+                                MatrixClientPeg.get().getRooms().find((r) => {
+                                    return r.getCanonicalAlias() === match[2];
+                                }) : MatrixClientPeg.get().getRoom(match[2]);
```

Comment:
Non-canonical aliases should surely work too? This will probably involve a server round-trip to look up the alias though :(